client/client.go

```diff
@@ -156,10 +156,123 @@ type tsoRequest struct {
 	dcLocation string
 }
 
+type tsoBatchController struct {
+	maxBatchSize         int
+	maxBatchWaitInterval time.Duration
+	// bestBatchSize is a dynamic size that changed based on the current batch effect.
+	bestBatchSize int
+
+	tsoRequestCh          chan *tsoRequest
+	collectedRequests     []*tsoRequest
+	collectedRequestCount int
+
+	batchStartTime time.Time
+}
+
+func newTSOBatchController(tsoRequestCh chan *tsoRequest, maxBatchSize int, maxBatchWaitInterval time.Duration) *tsoBatchController {
+	return &tsoBatchController{
+		maxBatchSize:          maxBatchSize,
+		maxBatchWaitInterval:  maxBatchWaitInterval,
+		bestBatchSize:         maxBatchSize,
+		tsoRequestCh:          tsoRequestCh,
+		collectedRequests:     make([]*tsoRequest, maxBatchSize+1),
+		collectedRequestCount: 0,
+	}
+}
+
+// fetchAllPendingRequests will start a new round of the batch collecting from the channel.
+// It returns true if everything goes well, otherwise false which should stop the service.
+func (tbc *tsoBatchController) fetchAllPendingRequests(ctx context.Context) bool {
+	var firstTSORequest *tsoRequest
+	select {
+	case <-ctx.Done():
+		return false
+	case firstTSORequest = <-tbc.tsoRequestCh:
+	}
+	// Start to batch when the first TSO request arrives.
+	tbc.batchStartTime = time.Now()
+	tbc.collectedRequestCount = 0
+	tbc.pushRequest(firstTSORequest)
+
+	// This loop is for trying best to collect more requests, so we use `tbc.maxBatchSize` here.
+	for tbc.collectedRequestCount < tbc.maxBatchSize {
```

Comment:
Will it affect the latency since we don't have a time limit here.

Reply:
BTW, do we need the handle the context in this for loop?

Reply:
I think it won't affect the latency since the `maxBatchSize` is not that large.

Reply:
If the `request` can be taken out all the time, it will be taken up to `maxBatchSize`. Otherwise, it will `break`. I don't think it will consume a lot of time. Of course, we can set a maximum value for `maxBatchSize`.